enterprise/wsproxy/wsproxysdk/wsproxysdk.go

```diff
@@ -204,6 +204,35 @@ type RegisterWorkspaceProxyRequest struct {
 	Version string `json:"version"`
 }
 
+type CryptoKeyFeature string
+
+const (
+	CryptoKeyFeatureWorkspaceApp  CryptoKeyFeature = "workspace_apps"
+	CryptoKeyFeatureOIDCConvert   CryptoKeyFeature = "oidc_convert"
+	CryptoKeyFeatureTailnetResume CryptoKeyFeature = "tailnet_resume"
+)
+
+type CryptoKey struct {
+	Feature   CryptoKeyFeature `json:"feature"`
+	Secret    string           `json:"secret"`
+	DeletesAt time.Time        `json:"deletes_at"`
+	Sequence  int32            `json:"sequence"`
+	StartsAt  time.Time        `json:"starts_at"`
+}
+
+func (c CryptoKey) Active(now time.Time) bool {
+	now = now.UTC()
+	isAfterStartsAt := !c.StartsAt.IsZero() && !now.Before(c.StartsAt)
+	return isAfterStartsAt && !c.Invalid(now)
+}
+
+func (c CryptoKey) Invalid(now time.Time) bool {
+	now = now.UTC()
+	noSecret := c.Secret == ""
+	afterDelete := !c.DeletesAt.IsZero() && !now.Before(c.DeletesAt.UTC())
+	return noSecret || afterDelete
+}
```

Comment:
Can this just be a single `Valid()` call or something instead? I don't understand why you would want to differentiate these two states since in both cases the key can't be used

Reply:
Keys can be used for verification prior to their start time but cannot be used for signing. I'll update the names of the methods to reflect that though